common/customizable_schema.go

```diff
@@ -13,6 +15,64 @@ type CustomizableSchema struct {
 	Schema         *schema.Schema
 	path           []string
 	isSuppressDiff bool
+	context        SchemaPathContext
+}
+
+type SchemaPathContext struct {
+	// Path is used for refernces from resourceProviderRegistry as prefix, follows the format of `[a,0,b]`
+	path       []string
+	schemaPath []*schema.Schema
+}
+
+func (spc SchemaPathContext) copy() SchemaPathContext {
+	newPath := make([]string, len(spc.path))
+	copy(newPath, spc.path)
+	newSchemaPath := make([]*schema.Schema, len(spc.schemaPath))
+	copy(newSchemaPath, spc.schemaPath)
+	return SchemaPathContext{
+		path:       newPath,
+		schemaPath: newSchemaPath,
+	}
+}
+
+func (spc SchemaPathContext) addToPath(fieldName string, schema *schema.Schema) SchemaPathContext {
+	newSpc := spc.copy()
+	// Special path element `"0"` is used to denote either arrays or sets of elements
+	newSpc.path = append(spc.path, fieldName, "0")
+	newSpc.schemaPath = append(spc.schemaPath, schema)
+	return newSpc
+}
+
+func getEmptySchemaPathContext() SchemaPathContext {
+	return SchemaPathContext{
+		[]string{},
+		[]*schema.Schema{},
+	}
+}
+
+func (s *CustomizableSchema) pathContainsMultipleItemsList() bool {
+	schemaPath := s.context.schemaPath
+	for _, scm := range schemaPath {
+		if scm.Type == schema.TypeList && scm.MaxItems != 1 {
+			return true
+		}
+	}
+	return false
+}
+
+// Used to get the prefix path for functions like ConflictsWith, by joining `path` in SchemaPathContext.
+func getPrefixedValue(path []string, value []string) []string {
+	var prefix string
+	if len(path) != 0 {
+		prefix = strings.Join(path, ".") + "."
+	} else {
+		prefix = ""
+	}
+	prefixedPaths := make([]string, len(value))
+	for i, item := range value {
+		prefixedPaths[i] = prefix + item
+	}
+	return prefixedPaths
 }
 
 func CustomizeSchemaPath(s map[string]*schema.Schema, path ...string) *CustomizableSchema {
```

Comment:
Can we remove this function?

Reply:
Not right now because there are already resources using them in the old `StructToSchema` callbacks